### DISCUSSION TITLE
Unpacking `TypeAliasType` to tuple of types using `*` errors at runtime, `Unpack` doesn't, pyright passes, bug in cpython or pyright?

### DISCUSSION THREAD

**Discussion by alwaysmpe** [microsoft/pyright#9343] [Q&A] [answered]
I'm writing classes using variadic generic code and as a minimal example I have the below code which raises no type errors on [pyright playground](https://pyright-play.net/?strict=true&code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwFgAoUSKJAT1y1IHNseAkQCqpXMQDGAawA0JevTSEYXXGhZq0UAGJgwAQRB8AKlAC8HBLgA2aANr0kIeU5ABdFi0k3ii3fr2AFRGfO4AXCxQ0VASil7MWgFgZpZ6YI7O8kHpoSaeidza6SKpyZkuUGISMva5xvkFCT5%2B9FAAQsQgIgAU6RXy1VLSdfp57u4AlJHMMbGtCQAmaMBQwPrwpD1IMF184ckm8si7xgclJtNRMa3KO1rbp3zy6ZfX0bcqDzt7L-qlk3eCiUXyKj1%2B5Tcg3Ew1GhgaE0BsxuijuqjBP2MfwyUKgOTGCKRc0%2B9zBJwhryJKJBpPUPXJWOSAKBJPRdIZz0hWSqMNq9VMiJZqNB7MxnP6uPx8IFSO8vn8nRAfUCkv5%2BSuyOicSYzCAA) and is crashing at runtime which I believe is a bug in cpython, would like some advice if possible. The code:

```python
from __future__ import annotations
from typing import Unpack, assert_type
type FooArgT = tuple[str, str]

class Foo[*Arg]:
    pass

type FooT = Foo[str, *FooArgT]
type FooUT = Foo[str, Unpack[FooArgT]]


class BarU(Foo[str, Unpack[FooArgT]]):
    pass

def foo_fn(t_arg: FooT, ut_arg: FooUT):
    assert_type(t_arg, FooT)
    assert_type(t_arg, FooUT)
    assert_type(t_arg, Foo[str, Unpack[FooArgT]])
    assert_type(t_arg, Foo[str, *FooArgT])
    assert_type(ut_arg, FooT)
    assert_type(ut_arg, FooUT)
    assert_type(ut_arg, Foo[str, Unpack[FooArgT]])
    assert_type(ut_arg, Foo[str, *FooArgT])

class Bar(Foo[str, *FooArgT]):
    pass
```

When I run this locally, I get:
```
Traceback (most recent call last):
  File "...", line 25, in <module>
    class Bar(Foo[str, *FooArgT]):
                  ^^^^^^^^^^^^^
TypeError: Value after * must be an iterable, not typing.TypeAliasType
```
By my understanding, the 2 class declarations are equivalent, pyright checks pass and cpython should allow unpacking of the aliased `TypeVarTuple`. A cursory look at the cpython issues I can't see anything equivalent so figured here might be a good place to start.

I'm running python3.13.0 in wsl installed using pyenv, output of `python -V -V`
```
Python 3.13.0 (main, Oct 24 2024, 00:12:39) [GCC 11.4.0]
```

Or should I just pack in this whole python thing and go back to c++?

**Comment by erictraut** [comment]
@JelleZijlstra, what do you think?  Should `TypeAliasType` work with an unpack operator? If not, it probably makes sense to add a check to static type checkers that reports an error if an unpack operator is applied to a `TypeAliasType` outside of a type annotation.

**Reply by JelleZijlstra** [reply, marked as answer]
I think we should make this work at runtime. @alwaysmpe could you open an issue on https://github.com/python/cpython about this?

**Reply by alwaysmpe** [reply]
Will do!

**Reply by alwaysmpe** [reply]
python/cpython#126085 created!
